File: sdk/ts/src/autopilot.ts
import { AgentOTC } from './client';
import { Deal } from './deal';

export interface AutoBuyerConfig {
    targetAsset: string;
    maxPrice: number;
    maxCollateral: number;
    /** Polling interval for new offers in milliseconds. Default: 60000 (1 min) */
    pollIntervalMs?: number;
    onMatch?: (deal: Deal) => Promise<void>;
    onSuccess?: (deal: Deal) => Promise<void>;
    onError?: (error: any) => Promise<void>;
}

export class AutoBuyer {
    private client: AgentOTC;
    private running: boolean = false;
    private config!: AutoBuyerConfig;
    private activeDealIds: Set<string> = new Set();
    private currentPollTimeout?: NodeJS.Timeout;

    constructor(client: AgentOTC) {
        this.client = client;
    }

    public async start(config: AutoBuyerConfig): Promise<void> {
        this.config = Object.assign({ pollIntervalMs: 60000 }, config);
        this.running = true;

        if (!this.client['ws'].isConnected) {
            await this.client.connect();
        }

        this.pollLoop();
    }

    public stop(): void {
        this.running = false;
        if (this.currentPollTimeout) {
            clearTimeout(this.currentPollTimeout);
        }
    }

    private async pollLoop(): Promise<void> {
        if (!this.running) return;

        try {
            const offers = await this.client.offers.list({ mode: 'sell', status: 'active', asset: this.config.targetAsset });
            
            for (const offer of offers) {
                if (offer.price <= this.config.maxPrice && offer.collateral <= this.config.maxCollateral) {
                    await this.attemptPurchase(offer.id);
                }
            }
        } catch (e) {
             if (this.config.onError) await this.config.onError(e);
        }

        if (this.running) {
            this.currentPollTimeout = setTimeout(() => this.pollLoop(), this.config.pollIntervalMs);
        }
    }

    private async attemptPurchase(offerId: string): Promise<void> {
        try {
            const deal = await this.client.offers.accept(offerId);
            
            if (this.activeDealIds.has(deal.id)) return;
            this.activeDealIds.add(deal.id);

            if (this.config.onMatch) {
                await this.config.onMatch(deal);
            }

            // Full Auto Execute Deal Lifecycle Workflow
            
            // 1. Agree to terms natively
            await deal.sendMessage(`@middleman I agree to purchase ${this.config.targetAsset} at ${this.config.maxPrice} SOL.`);
            
            // 2. Wait for Escrow
            await deal.waitForPhase(['escrow_created', 'awaiting_deposits'], { timeoutMs: 120000 });
            
            // 3. Deposit collateral locally
            await deal.depositToEscrow(this.config.maxCollateral, 'buyer');
            
            // 4. Wait for Phase: delivery
            await deal.waitForPhase('delivery', { timeoutMs: 120000 });

            // 5. Send Payment
            await deal.depositToEscrow(this.config.maxPrice, 'buyer');

            // 6. Confirm receipt and completion
            await deal.confirmDelivery();
            await deal.waitForPhase('completed', { timeoutMs: 120000 });

            if (this.config.onSuccess) {
                await this.config.onSuccess(deal);
            }
        } catch (e: any) {
             if (this.config.onError) await this.config.onError(e);
        }
    }
}

File: sdk/python/agentotc/src/agentotc/autopilot.py
import asyncio
import logging
from typing import Callable, Optional, Set, Coroutine

from .client import AgentOTC
from .deal import Deal

logger = logging.getLogger("AgentOTC.AutoBuyer")

class AutoBuyer:
    def __init__(self, client: AgentOTC):
        self.client = client
        self.running = False
        self.active_deal_ids: Set[str] = set()
        
        self.target_asset: str = ""
        self.max_price: float = 0.0
        self.max_collateral: float = 0.0
        self.poll_interval_ms: int = 60000
        
        self.on_match: Optional[Callable[[Deal], Coroutine]] = None
        self.on_success: Optional[Callable[[Deal], Coroutine]] = None
        self.on_error: Optional[Callable[[Exception], Coroutine]] = None

    async def start(self, 
                    target_asset: str, 
                    max_price: float, 
                    max_collateral: float, 
                    poll_interval_ms: int = 60000,
                    on_match: Callable[[Deal], Coroutine] = None,
                    on_success: Callable[[Deal], Coroutine] = None,
                    on_error: Callable[[Exception], Coroutine] = None):
        self.target_asset = target_asset
        self.max_price = max_price
        self.max_collateral = max_collateral
        self.poll_interval_ms = poll_interval_ms
        self.on_match = on_match
        self.on_success = on_success
        self.on_error = on_error
        
        self.running = True

        if not self.client.ws.is_connected:
            await self.client.connect()

        asyncio.create_task(self._poll_loop())

    def stop(self):
        self.running = False

    async def _poll_loop(self):
        while self.running:
            try:
                offers = await self.client.offers.list(asset=self.target_asset, mode='sell', status='active')
                
                for offer in offers:
                    if offer.price <= self.max_price and offer.collateral <= self.max_collateral:
                        # Non-blocking attempt
                        asyncio.create_task(self._attempt_purchase(offer.id))
                        
            except Exception as e:
                if self.on_error:
                    await self.on_error(e)
                else:
                    logger.error(f"[AutoBuyer] Loop error: {e}")
            
            await asyncio.sleep(self.poll_interval_ms / 1000.0)

    async def _attempt_purchase(self, offer_id: str):
        try:
            deal = await self.client.offers.accept(offer_id)
            
            if deal.id in self.active_deal_ids:
                return
            self.active_deal_ids.add(deal.id)

            if self.on_match:
                await self.on_match(deal)

            # 1. Agree
            await deal.send_message(f"@middleman I agree to purchase {self.target_asset} at {self.max_price} SOL.")
            
            # 2. Escrow
            await deal.wait_for_phase('wait_escrow')
            await deal.deposit_to_escrow(self.max_collateral, 'buyer')
            
            # 3. Delivery
            await deal.wait_for_phase('wait_delivery')
            await deal.deposit_to_escrow(self.max_price, 'buyer')

            # 4. Confirm
            await deal.confirm_delivery()
            await deal.wait_for_phase('completed')

            if self.on_success:
                await self.on_success(deal)

        except Exception as e:
            if self.on_error:
                await self.on_error(e)
            else:
                logger.error(f"[AutoBuyer] Deal {offer_id} failed: {e}")
